docs/cudf/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,268 @@
+  A column that is not a slice has an offset of 0.
+
+More information about these fields can be found in the documentation of the
+[Apache Arrow Columnar Format](https://arrow.apache.org/docs/format/Columnar.html),
+which is what the cuDF `Column` is based on.
+
+The `Column` class is implemented in Cython to facilitate interoperability with `libcudf`'s C++ data structures.
+Most higher-level functionality is implemented in the `ColumnBase` subclass.
+These functions rely `Column` APIs to call `libcudf` APIs and translate their results to Python.
+This separation allows `ColumnBase` to be implemented in pure Python, which simplifies development and debugging.
+
+`ColumnBase` provides some standard methods, while other methods only make sense for data of a specific type.
+As a result, we have various subclasses of `ColumnBase` like `NumericalColumn`, `StringColumn`, and `DatetimeColumn`.
+Most dtype-specific decisions should be handled at the level of a specific `Column` subclass.
+Each type of `Column` only implements methods supported by that data type.
+
+Different types of `ColumnBase` are also stored differently in memory according to the Arrow format.
+As one example, a `NumericalColumn` with 1000 `int32` elements and containing nulls is composed of:
+
+1. A data buffer of size 4000 bytes (sizeof(int32) * 1000)
+2. A mask buffer of size 128 bytes (1000/8 padded to a multiple of 64
+   bytes)
+3. No children columns
+
+As another example, a `StringColumn` backing the Series `['do', 'you', 'have', 'any', 'cheese?']` is composed of:
+
+1. No data buffer
+2. No mask buffer as there are no nulls in the Series
+3. Two children columns:
+
+   - A column of UTF-8 characters
+     `['d', 'o', 'y', 'o', 'u', 'h', ..., '?']`
+   - A column of "offsets" to the characters column (in this case,
+     `[0, 2, 5, 9, 12, 19]`)
+
+
+### Data types
+
+cuDF uses [dtypes](https://numpy.org/doc/stable/reference/arrays.dtypes.html) to represent different types of data.
+Since efficient GPU algorithms require preexisting knowledge of data layouts,
+cuDF does not support the arbitrary `object` dtype, but instead defines a few custom types for common use-cases:
+- `ListDtype`: Lists where each element in every list in a Column is of the same type
+- `StructDtype`: Dicts where a given key always maps to values of the same type
+- `CategoricalDtype`: Analogous to the pandas categorical dtype except that the categories are stored in device memory
+- `DecimalDtype`: Fixed-point numbers
+- `IntervalDtype`: Intervals
+
+Note that there is a many-to-one mapping between data types and `Column` classes.
+For instance, all numerical types (floats and ints of different widths) are all managed using `NumericalColumn`.
+
+
+### Buffer
+
+Although a `Column` represents an Arrow-compliant data structure, it does not directly handle memory mangaement.
+That job is delegated to the `Buffer` class, which represents a device memory allocation owned by another object.
```

Comment:
I think thinking of buffers as "memory managers" is maybe less useful than thinking of Columns as being composed of one or more buffers:

```suggestion
`Column`s are in turn composed of one or more `Buffer`s.
A `Buffer` represents a single, contiguous, device memory allocation owned by another object.
```